### ISSUE TITLE
Writing to overlay returns "No Data Available"

### ISSUE THREAD

**Issue by bdwheele** [apptainer/apptainer#1387]
### Version of Apptainer

```
$ apptainer --version
apptainer version 1.1.8-1.el8

$ apptainer --version
apptainer version 1.1.7-1.fc38
```

### Expected behavior
Overlay of sif should allow writes to normally non-writable portions of the filesystem (such as /usr/local) as it did when I was
running Singularity 3.8.5. 

### Actual behavior
Writes to overlaid data returns -ENODATA.  

### Steps to reproduce this behavior
On RockyLinux 8.8:
```
$ cat > deb10.recipe
Bootstrap: docker
From: debian:10
$ apptainer build deb10.sif deb10.recipe
...
$ apptainer overlay create --size 100 deb10.overlay
$ apptainer run --overlay deb10.overlay deb10.sif
INFO:    unknown argument ignored: xino=on
/proc seems to be mounted as readonly, it can lead to unexpected failures
Apptainer> touch /usr/local/foo
touch: cannot touch '/usr/local/foo': No data available
Apptainer>
```
On Fedora 38: 
```
... all the above commands and output are the same ...
$ apptainer run --overlay deb10.overlay deb10.sif
Apptainer> touch /usr/local/foo
touch: cannot touch '/usr/local/foo': No data available
```

### What OS/distro are you running

```
$ cat /etc/rocky-release
Rocky Linux release 8.8 (Green Obsidian)

$ cat /etc/fedora-release
Fedora release 38 (Thirty Eight)
```

### How did you install Apptainer
RPMs:
* apptainer-1.1.7-1.fc38.x86_64  (on Fedora 38)
* apptainer-1.1.8-1.el8.x86_64 (on RockyLinux 8.8)

**Comment by narein257** [comment]
I have a similar problem,. I have attached the outputs from my Rocky 8.7.

x-special/nautilus-clipboard
copy
file:///home/lnarayanan/Documents/log.txt

**Comment by bdwheele** [comment]
@DrDaveD -  correct, this is unprivileged and the suid package is not installed.  I don't believe I've modified any settings in /etc/apptainer so it should be a bog-standard installation.

I dug around a bit and I don't see any messages in the host's system log or the selinux audit log.  The ENODATA seems to be coming from the fuse overlay filesystem as near as I can tell.

Here's the kernel version and the version for fuse-overlayfs:
```
Linux localhost.localdomain 4.18.0-477.13.1.el8_8.x86_64 #1 SMP Tue May 30 22:15:39 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
fusermount3 version: 3.3.0
fuse-overlayfs: version 1.11
FUSE library version 3.3.0
using FUSE kernel interface version 7.26
```
(fuse-overlayfs does drop a line to stderr about not knowing the argument "lazytime", not sure where that's coming from)



This is the debug output for the container described above just running "touch /usr/local/foo".

[debug.txt](https://github.com/apptainer/apptainer/files/11723035/debug.txt)

**Comment by DrDaveD** [comment]
@narein257 You don't show your command line here, I assume it is the same as in #1362, but do you still get the same error by simplifying your command (e.g. without doing any of the network options -- let me know what you try)?  

@bdwheele Can you reproduce the problem with `--writable-tmpfs` instead of `--overlay`?  I also don't know where that "lazytime" option is coming from. Apptainer doesn't set it and I don't get that error.  I am also using fuse-overlayfs version 1.11, the latest version available in Rocky8 appstream.  I don't see the message in your debug.txt output although I do see it in the log.txt output from @narein257.  I'm wondering if that option is significant, especially since you are both getting it and I am not.  The option means to delay writing timestamps.  Have you perhaps set it in /etc/fstab for one of your filesystems?

**Comment by DrDaveD** [comment]
And another random idea -- do you have SELinux enabled, and if so could it be some privilege violation there?  I have it disabled.

**Comment by bdwheele** [comment]
I just tried it with writable-tmpfs:
```
$ apptainer exec --writable-tmpfs deb10.sif touch /usr/local/foo
INFO:    unknown argument ignored: xino=on
/proc seems to be mounted as readonly, it can lead to unexpected failures
/usr/bin/touch: cannot touch '/usr/local/foo': No data available
```

I've got selinux enabled (and it's enforcing), but I'm not seeing anything in /var/log/audit/audit.log.

I did a grep for 'lazytime' and it's in /usr/bin/apptainer, /usr/bin/fuse-overlayfs, and /usr/libexec/apptainer/bin/starter...all of which are binaries.

**Comment by DrDaveD** [comment]
I enabled selinux and it didn't help me reproduce the problem.

It's interesting that you see the "No data available" error also with `--writable-tmpfs` because that's using a different upper layer type.   This leads me to consider more carefully the lower layer, which is squashfuse_ll, or fuse-overlayfs itself.  The rpm version of fuse-overlayfs is `fuse-overlayfs-1.11-1.module+el8.8.0+1265+fa25dd7a.x86_64`; does that match what you have? squashfuse_ll is coming from the apptainer rpm for both of us.  The next thing I would probably do if I could reproduce this would be to replace the fuse-overlayfs command with a script that either enables debugging in that layer or runs strace on it.

Yes apptainer includes the "lazytime" string but it never sets the option, it only recognizes the mount option.  You can see that if you search the source code.  Interestingly, I do not see the "lazytime" warning when you used `--writable-tmpfs`; is it true, that's only there when you use `--overlay`?  If so it appears to be unrelated to the underlying "No data available" since you get that with or without the "lazytime" warning.  (The xino=on and /proc warnings are ignored in version 1.1.9).

**Comment by bdwheele** [comment]
Yes, I have the same version of fuse-overlayfs.

Here's a script that I'm running on Fedora 38 that uses qemu-7.2.1-1.fc38.x86_64 to build a VM that can reproduce the issue.  I also reproduced it in a home-built 7.0.0 qemu running on RockyLinux 8.7.  It'll probably work on other distros and versions of qemu as well.

[runme.sh.txt](https://github.com/apptainer/apptainer/files/11760000/runme.sh.txt)

After the install is complete, reboot the VM and log in as testuser (password is in the script).   I was able to reproduce it:

![Screenshot_2023-06-15_11-36-57](https://github.com/apptainer/apptainer/assets/3912076/9a82e7ae-5007-42cd-bc59-a68281d977df)

Hopefully that will give you an environment that exhibits the same issue.  If not...well, I have no idea what that would mean

**Comment by narein257** [comment]
Here is the command I used: apptainer -d run --fakeroot --net --network-args="portmap=8080:80/tcp" --overlay /home/lnarayanan/Documents/tmpfs/ nginx.sif &> log.txt
I get the same beviour even when not using --net commands: apptainer run --fakeroot --overlay /home/lnarayanan/Documents/tmpfs/ nginx.sif

**Comment by DrDaveD** [comment]
That sounds very promising!  After much fiddling (including recompilng qemu from source rpm to enable `vvfat` format) I was able to get qemu to launch with this script (only modified to run qemu-kvm instead of qemu-system-x86_64) on an el7 host, and connect a VNC viewer to it.  It gave boot prompts and by default went into a standard Rocky Linux installation, which didn't look right because it right away asks for a root password and a user account.   I was able to boot into rescue mode, mount the vfat partition /dev/sdb1, and see ks.cfg. Then following Red Hat documentation I found that I needed to add the boot option "init.ks=hd:sdb1:/ks.cfg".  I did that by hitting "<tab>" on one of the boot options, then it gave the kernel boot parameters and it showed the automation of the installation process.  Then after a long time I was indeed able to log in and reproduce the issue.  Thanks!  Now I will proceed to try to debug it.

**Comment by bdwheele** [comment]
Glad you were able to reproduce it!  Sorry it was so problematic.

**Comment by bdwheele** [comment]
I was trying to test it -- but I'm not sure how?  Is there a flag in the build command (or elsewhere) that adds the --no-xattrs somewhere?

**Comment by DrDaveD** [comment]
There is not a build option for that, it would need to modify the place where apptainer invokes mksquashfs.

However, I have since found the real root cause, a bug in squashfuse, and have an apptainer fix in #1478.
